packages/stylelint-config-scss/index.js

```diff
@@ -165,6 +165,15 @@ module.exports = {
     'csstools/use-nesting': 'always',
 
     // use valid css
-    'csstree/validator': true,
+    'csstree/validator': {
+      'syntaxExtensions': ['sass'],
+      'ignoreAtrules': ['else'],
```

Comment:
Can you please leave a comment on why do we ignore `@else`?

Reply:
If the `else`at rule isn't ignored, the following error keeps popping up.  `At-rule '@else' should contain a prelude (csstree/validator)`. For example, in `stylelint-config-scss › feature if-else-loop › correct style › flags no warnings`

I do not quite know for what reason this error is prompted or how to mitigate it. Could you assist?

Reply:
Can you please add a reproducer?

Reply:
Unable to reproduce via an online platform like Codesandbox.

1. In `packages/stylelint-config-scss/index.js`, remove `'ignoreAtrules': ['else']`
2. `yarn test`
3. Result:

```sh
yarn run v1.22.11
$ yarn run test:unit && yarn run lint
$ lerna run test
lerna notice cli v3.22.1
lerna info versioning independent
lerna info Executing command in 3 packages: "yarn run test"
lerna ERR! yarn run test exited 1 in '@wemake-services/stylelint-config-scss'
lerna ERR! yarn run test stdout:
$ jest
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

lerna ERR! yarn run test stderr:
FAIL @wemake-services/stylelint-config-scss tests/index.spec.js (17.151 s)
  stylelint-config-scss
    feature declarations
      correct style
        ✓ flags no warnings (8266 ms)
      incorrect style
        ✓ flags correct number of warnings (33 ms)
        ✓ flags correct rule warnings (2 ms)
    feature functions
      correct style
        ✓ flags no warnings (29 ms)
      incorrect style
        ✓ flags correct number of warnings (27 ms)
        ✓ flags correct rule warnings (1 ms)
    feature if-else-loop
      correct style
        ✕ flags no warnings (29 ms)
      incorrect style
        ✕ flags correct number of warnings (23 ms)
        ✓ flags correct rule warnings (3 ms)
    feature imports
      correct style
        ✓ flags no warnings (17 ms)
      incorrect style
        ✓ flags correct number of warnings (13 ms)
        ✓ flags correct rule warnings
    feature media-queries
      correct style
        ✓ flags no warnings (29 ms)
      incorrect style
        ✓ flags correct number of warnings (23 ms)
        ✓ flags correct rule warnings (1 ms)
    feature mixins
      correct style
        ✓ flags no warnings (20 ms)
      incorrect style
        ✓ flags correct number of warnings (19 ms)
        ✓ flags correct rule warnings
    feature operators
      correct style
        ✓ flags no warnings (15 ms)
      incorrect style
        ✓ flags correct number of warnings (16 ms)
        ✓ flags correct rule warnings (1 ms)
    feature order
      correct style
        ✓ flags no warnings (26 ms)
    feature placeholders
      correct style
        ✓ flags no warnings (15 ms)
      incorrect style
        ✓ flags correct number of warnings (17 ms)
        ✓ flags correct rule warnings
    feature plugins
      correct style
        ✓ flags no warnings (15 ms)
      incorrect style
        ✓ flags correct number of warnings (17 ms)
        ✓ flags correct rule warnings (1 ms)
    feature selectors
      correct style
        ✓ flags no warnings (34 ms)
      incorrect style
        ✓ flags correct number of warnings (35 ms)
        ✓ flags correct rule warnings
    feature url
      correct style
        ✓ flags no warnings (15 ms)
      incorrect style
        ✓ flags correct number of warnings (15 ms)
        ✓ flags correct rule warnings (1 ms)
    feature variables
      correct style
        ✓ flags no warnings (13 ms)
      incorrect style
        ✓ flags correct number of warnings (13 ms)
        ✓ flags correct rule warnings (1 ms)
    feature vendor
      correct style
        ✓ flags no warnings (23 ms)
      incorrect style
        ✓ flags correct number of warnings (20 ms)
        ✓ flags correct rule warnings

  ● stylelint-config-scss › feature if-else-loop › correct style › flags no warnings

    expect(received).toHaveLength(expected)

    Expected length: 0
    Received length: 3
    Received array:  [{"column": 1, "line": 13, "rule": "scss/at-mixin-parentheses-space-before", "severity": "error", "text": "Expected a single space before parentheses in mixin declaration (scss/at-mixin-parentheses-space-before)"}, {"column": 1, "line": 1, "rule": "scss/at-function-parentheses-space-before", "severity": "error", "text": "Expected a single space before parentheses in function declaration (scss/at-function-parentheses-space-before)"}, {"column": 11, "line": 8, "rule": "csstree/validator", "severity": "error", "text": "At-rule `@else` should contain a prelude (csstree/validator)"}]

      175 |
      176 |       const lintData = await lintResult
    > 177 |       expect(lintData.results[0].warnings).toHaveLength(0)
          |                                            ^
      178 |       expect(lintData.errored).toBe(false)
      179 |     }, 15000) // some tasks are executed for a long time
      180 |   })

      at _callee$ (packages/stylelint-config-scss/tests/index.spec.js:177:44)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:45:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:274:22)
      at Generator.prototype.<computed> [as next] (node_modules/regenerator-runtime/runtime.js:97:21)
      at asyncGeneratorStep (packages/stylelint-config-scss/tests/index.spec.js:19:103)
      at _next (packages/stylelint-config-scss/tests/index.spec.js:21:194)

  ● stylelint-config-scss › feature if-else-loop › correct style › flags no warnings

    expect.assertions(2)

    Expected two assertions to be called but received one assertion call.

      172 |     test('flags no warnings', async () => {
      173 |       expect.hasAssertions()
    > 174 |       expect.assertions(2)
          |              ^
      175 |
      176 |       const lintData = await lintResult
      177 |       expect(lintData.results[0].warnings).toHaveLength(0)

      at _callee$ (packages/stylelint-config-scss/tests/index.spec.js:174:14)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:45:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:274:22)
      at Generator.prototype.<computed> [as next] (node_modules/regenerator-runtime/runtime.js:97:21)
      at asyncGeneratorStep (packages/stylelint-config-scss/tests/index.spec.js:19:103)
      at _next (packages/stylelint-config-scss/tests/index.spec.js:21:194)
      at packages/stylelint-config-scss/tests/index.spec.js:21:364
      at Object.<anonymous> (packages/stylelint-config-scss/tests/index.spec.js:21:97)

  ● stylelint-config-scss › feature if-else-loop › incorrect style › flags correct number of warnings

    expect(received).toHaveLength(expected)

    Expected length: 13
    Received length: 14
    Received array:  [{"column": 4, "line": 9, "rule": "scss/at-else-closing-brace-newline-after", "severity": "error", "text": "Unexpected newline after \"}\" of @else statement (scss/at-else-closing-brace-newline-after)"}, {"column": 4, "line": 9, "rule": "scss/at-else-closing-brace-space-after", "severity": "error", "text": "Expected single space after \"}\" of @else statement (scss/at-else-closing-brace-space-after)"}, {"column": 4, "line": 7, "rule": "scss/at-else-if-parentheses-space-before", "severity": "error", "text": "Expected a single space before parentheses in else-if declaration (scss/at-else-if-parentheses-space-before)"}, {"column": 4, "line": 7, "rule": "scss/at-if-closing-brace-space-after", "severity": "error", "text": "Expected single space after \"}\" of @if statement (scss/at-if-closing-brace-space-after)"}, {"column": 9, "line": 10, "rule": "csstree/validator", "severity": "error", "text": "At-rule `@else` should contain a prelude (csstree/validator)"}, {"column": 3, "line": 4, "rule": "at-rule-empty-line-before", "severity": "error", "text": "Expected empty line before at-rule (at-rule-empty-line-before)"}, {"column": 5, "line": 4, "rule": "at-rule-name-space-after", "severity": "error", "text": "Expected single space after at-rule name \"@if\" (at-rule-name-space-after)"}, {"column": 23, "line": 10, "rule": "block-closing-brace-space-before", "severity": "error", "text": "Expected single space before \"}\" of a single-line block (block-closing-brace-space-before)"}, {"column": 9, "line": 10, "rule": "block-opening-brace-space-after", "severity": "error", "text": "Expected single space after \"{\" of a single-line block (block-opening-brace-space-after)"}, {"column": 2, "line": 5, "rule": "block-opening-brace-space-before", "severity": "error", "text": "Expected single space before \"{\" (block-opening-brace-space-before)"}, {"column": 7, "line": 10, "rule": "block-opening-brace-space-before", "severity": "error", "text": "Expected single space before \"{\" (block-opening-brace-space-before)"}, {"column": 2, "line": 15, "rule": "block-opening-brace-space-before", "severity": "error", "text": "Expected single space before \"{\" (block-opening-brace-space-before)"}, {"column": 25, "line": 22, "rule": "block-opening-brace-space-before", "severity": "error", "text": "Expected single space before \"{\" (block-opening-brace-space-before)"}, {"column": 1, "line": 1, "rule": "no-empty-first-line", "severity": "error", "text": "Unexpected empty line (no-empty-first-line)"}]

      195 |       const lintData = await lintResult
      196 |
    > 197 |       expect(lintData.results[0].warnings).toHaveLength(rule.errors)
          |                                            ^
      198 |       expect(lintData.errored).toBe(true)
      199 |     })
      200 |

      at _callee2$ (packages/stylelint-config-scss/tests/index.spec.js:197:44)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:45:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:274:22)
      at Generator.prototype.<computed> [as next] (node_modules/regenerator-runtime/runtime.js:97:21)
      at asyncGeneratorStep (packages/stylelint-config-scss/tests/index.spec.js:19:103)
      at _next (packages/stylelint-config-scss/tests/index.spec.js:21:194)

  ● stylelint-config-scss › feature if-else-loop › incorrect style › flags correct number of warnings

    expect.assertions(2)

    Expected two assertions to be called but received one assertion call.

      191 |     test('flags correct number of warnings', async () => {
      192 |       expect.hasAssertions()
    > 193 |       expect.assertions(2)
          |              ^
      194 |
      195 |       const lintData = await lintResult
      196 |

      at _callee2$ (packages/stylelint-config-scss/tests/index.spec.js:193:14)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:45:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:274:22)
      at Generator.prototype.<computed> [as next] (node_modules/regenerator-runtime/runtime.js:97:21)
      at asyncGeneratorStep (packages/stylelint-config-scss/tests/index.spec.js:19:103)
      at _next (packages/stylelint-config-scss/tests/index.spec.js:21:194)
      at packages/stylelint-config-scss/tests/index.spec.js:21:364
      at Object.<anonymous> (packages/stylelint-config-scss/tests/index.spec.js:21:97)

Test Suites: 1 failed, 1 total
Tests:       2 failed, 38 passed, 40 total
Snapshots:   0 total
Time:        17.639 s
Ran all test suites.
error Command failed with exit code 1.

lerna ERR! yarn run test exited 1 in '@wemake-services/stylelint-config-scss'
lerna WARN complete Waiting for 2 child processes to exit. CTRL-C to exit immediately.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Is this sufficient information?

Reply:
Can you please try setting explicit`'scss/at-mixin-parentheses-space-before': 'always'`? https://github.com/stylelint-scss/stylelint-scss/blob/master/src/rules/at-mixin-parentheses-space-before/README.md

Reply:
This rule is already explicitly set on line 104.
https://github.com/wemake-services/wemake-frontend-styleguide/blob/700e8bc3498ab43be996df6e66d4a90a552276c8/packages/stylelint-config-scss/index.js#L104

Reply:
I think it's an error in `stylelint-validator` dependency.

https://github.com/csstree/stylelint-validator/blob/7897344972df87e1fe3ebedd35f1932b292242ac/lib/syntax-extension/sass/index.js#L37-L39

If we append `@else` with brackets, `@else ()`, the validator does not prompt the `At-rule '@else' should contain a prelude (csstree/validator)` error.

Reply:
I fell into the `csstree` rabbit hole trying to fix this error, without luck. I'm unable to fix correctly validating `@else` rules quickly. Particularly, the combination with `@else if` creates a weird edge case, as the csstree `at-rule` name declarations does not support spacing characters and treats the `if` of the `@else if` as a prelude.

I propose ignoring this at-rule for this PR. When I have more time available, I will open an issue / PR in the `stylelint-validator` repo.